### PR TITLE
drivers: sensor: Add tmd2620

### PR DESCRIPTION
**Pull request by pefech**
This PR adds the driver for the IR proximity sensor tmd2620 made by ams-osram, including a sample.
The driver works in polling and trigger mode with power management capabilities and its possible to configure almost every setting of the sensor via Kconfig / Menuconfig. Every working mode can be tested with the sample.

You can find datasheets of the sensor here: [TMD2620 datasheet](https://ams.com/documents/20143/36005/TMD2620_DS000461_3-00.pdf)

### CODE REVIEW

File: drivers/sensor/tmd2620/Kconfig [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		2: 5.56ms
+		...
+	default 255
+	range 1 256 # decrement by 1
+
+config TMD2620_WLONG
+	bool "Wait Long"
+	help
+		increase wait time by a factor of 12x
+	default n
```
Comment (MaureenHelm) [inline drivers/sensor/tmd2620/Kconfig:119]: Please convert these to devicetree properties to allow per-instance configuration

File: drivers/sensor/tmd2620/tmd2620.c [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+	.attr_set = tmd2620_attr_set,
+	.trigger_set = tmd2620_trigger_set,
+#endif
+};
+
+PM_DEVICE_DT_INST_DEFINE(0, tmd2620_pm_action);
+
+DEVICE_DT_INST_DEFINE(0, tmd2620_init, PM_DEVICE_DT_INST_GET(0),
+			&tmd2620_data, &tmd2620_config, POST_KERNEL,
+			CONFIG_SENSOR_INIT_PRIORITY, &tmd2620_driver_api);
```
Comment (MaureenHelm) [inline drivers/sensor/tmd2620/tmd2620.c:441]: Please use `DT_INST_FOREACH_STATUS_OKAY` to support multiple instances

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thank you for the contribution.

Please add this sensor to `tests/drivers/build_all/sensor/` and modify the sample to be independent of the specific driver (i.e., a generic proximity sensor sample).

There also seems to be an overabundance of tabs to indent lines. You should configure your text editor indent size to 8.

**Comment by pefech** [comment]
Hi, thank you for reviewing so quickly!
I made the driver support multiple instances and changed the configuration for the sensor settings to be configured in the device tree.
But I'm struggling a little bit with writing a generic proximity sensor sample, because the most drivers use Kconfig to be configured in trigger or polling mode (like the tmd2620 driver too). I'm not sure how to implement a statement into the sample that is able to detect if the driver has to be used in trigger or polling mode because the Kconfig entries are all diffrent. For example the apds9960 uses ```CONFIG_APDS9960_TRIGGER``` and my implementation for the tmd2620 uses ```CONFIG_TMD2620_TRIGGER```.

The only option that comes to my mind would be to have two samples - one for polling and one for trigger mode.
Or do I miss something important?

**Comment by pefech** [comment]
The last push includes all your requested changes.

File: samples/sensor/proximity_polling/src/main.c [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	while (true) {
+		k_sleep(K_MSEC(2000));
+		print_prox_data();
+
+#ifdef CONFIG_PM_DEVICE
+		pm_device_action_run(dev, PM_DEVICE_ACTION_SUSPEND);
+		printk("set low power state for 2s\n");
+		k_sleep(K_MSEC(2000));
+		pm_device_action_run(dev, PM_DEVICE_ACTION_RESUME);
+#endif
```
Comment (gmarull) [inline samples/sensor/proximity_polling/src/main.c:59]: please don't use this API in applications, you should be using runtime PM framework instead.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: samples/sensor/proximity_polling/sample.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  sample.sensor.proximity_polling:
+    harness: console
+    tags: sensors proximity
+    filter: dt_alias_exists("proximity-sensor0")
+    harness_config:
+      type: one_line
+      regex:
+        - "Proximity sensor sample application"
+        - "Found [0-9] proximity sensor(s): (.*)$"
+        - "Proximity on ([^:]*): [0-9]*$"
```
[the remark this answers is not in the document]
Comment (pefech) [reply samples/sensor/proximity_polling/sample.yaml:13]: If I understood it correctly the proximity value is an adimensional value from 1 to 255 where 1 is near and 255 is far.
It has been described here like this: [sensor_channel](https://docs.zephyrproject.org/apidoc/latest/group__sensor__interface.html#gaaa1b502bc029b10d7b23b0a25ef4e934)

It also depends on the settings of the sensor itself (and maybe the material it is supposed to detect) how far an objects proximity would be converted to a unit.

File: dts/bindings/sensor/ams,tmd2620.yaml [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        0-> every cycle fires an interrupt
+        1-> 1 consecutive proximity value out of threshold range fires an interrupt
+        2-> 2 consecutive proximity values out of threshold range fires an interrupt
+        ...
+
+    enable-wait-mode:
+      type: boolean
+      description: Enables wait mode
+
+    wait-time:
```
[the remark this answers is not in the document]
Comment (pefech) [reply dts/bindings/sensor/ams,tmd2620.yaml:76]: the units is specified in the description of ```wait-time```. I'll rename it to ```wait-time-factor```.

**Pull request closed without merging**